### PR TITLE
raftstore: reduce message flush

### PR DESCRIPTION
**Pull request by BusyJay**
###  What have you changed?

Technically, a thread only needs to flush messages when it's going to
block. This PR add hooks to every block point instead of flushing at
the end of every ready.

###  What is the type of the changes?

- Improvement

###  How is the PR tested?

integration tests

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No.

###  Does this PR affect `tidb-ansible`?

No.

###  Benchmark result if necessary (optional)

Benchmark shows that when there are about 25k idle regions, raftstore CPU is reduced from 55% to 40%, and grpc is reduced from 60% to 40%。

### CODE REVIEW

**Comment by siddontang** [comment]
amazing @BusyJay 

> Benchmark shows that when there are about 25k idle regions, raftstore CPU is reduced from 55% to 40%, and grpc is reduced from 60% to 40%。

Is it tested without Hibernate region?

**Comment by BusyJay** [comment]
Yes.

**Comment by overvenus** [comment]
> This PR add hooks to every block point instead of flushing at the end of every ready.

Sounds like a large batch may increase commit duration? 🤔

**Comment by siddontang** [comment]
maybe we can use commit duration metric to verify this.

File: src/raftstore/store/fsm/batch.rs [hunk 1]
```diff
@@ -252,6 +252,9 @@ pub trait PollHandler<N, C> {
 
     /// This function is called at the end of every round.
     fn end(&mut self, batch: &mut [Box<N>]);
+
+    /// This function is called when batch system is going to sleep.
+    fn pause(&mut self) {}
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/fsm/batch.rs]: I prefer name it yield.

**Comment by siddontang** [comment]
how about if you enable hibernate region?

File: src/raftstore/store/fsm/batch.rs (hunk 1 above)
Comment (BusyJay) [reply src/raftstore/store/fsm/batch.rs]: Why? `PollHandler` can't do the yield job. And in my opinion, yield usually means short switch.

**Comment by BusyJay** [comment]
> > This PR add hooks to every block point instead of flushing at the end of every ready.
> 
> Sounds like a large batch may increase commit duration? 🤔

No, raft client will flush internally every time it receives 8 messages.

**Comment by siddontang** [comment]
>  it receives 8 messages

why 8 messages? can we configure it?

**Comment by siddontang** [comment]
seem the performance not increased, interesting.

**Comment by zhangjinpeng1987** [comment]
> seem the performance not increased, interesting.

Maybe the dataset is large enough.

**Comment by BusyJay** [comment]
> why 8 messages? can we configure it?

It's hard coded by batch raft. Further plan is to refactor the whole batch mechanism.

**Comment by BusyJay** [comment]
PTAL

File: src/raftstore/store/fsm/store.rs [hunk 2]
```diff
@@ -471,7 +471,9 @@ impl<T: Transport, C: PdClient> RaftPoller<T, C> {
                     .schedule_task(prop.region_id, ApplyTask::Proposal(prop));
             }
         }
-        if self.poll_ctx.need_flush_trans {
+        if self.poll_ctx.need_flush_trans
+            && (!self.poll_ctx.kv_wb.is_empty() || !self.poll_ctx.raft_wb.is_empty())
+        {
             self.poll_ctx.trans.flush();
```
Comment (hicqu) [inline src/raftstore/store/fsm/store.rs]: Seems messages are flushed before disk write. If the messsages contain vote messages, could the peer vote twice in a term if the machine fail after here and before disk write?

Comment (BusyJay) [reply src/raftstore/store/fsm/store.rs]: If there are pending messages, it means they are from leader, it's OK to flush before writing disk. Also related to pingcap/raft-rs#292.

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**